src/Data/Vector/NonEmpty/Internal.hs

```diff
@@ -85,6 +89,21 @@ instance Read1 NonEmptyVector where
 instance Foldable NonEmptyVector where
     foldMap f = Foldable.foldMap f . _neVec
 
+#if MIN_VERSION_base(4,18,0)
+instance Foldable1 NonEmptyVector where
+    foldMap1 f vne =
+          let x = V.unsafeHead $ _neVec vne
+              xs = V.unsafeTail $ _neVec vne
+          in  go (f x) xs
+      where go y vec =
+              let z = V.unsafeHead vec
+                  zs = V.unsafeTail vec
+              in if V.null zs then
+                  y <> (f z)
+                else
+                  y <> go (f z) zs
```

Comment:
This is going to explode if there is only 1 element in the list. Usually, implementations special case the length 1  in order to avoid this blow up. Failing example: 

```haskell
П> foldMap Sum $ unsafeFromList [1]
Sum {getSum = 1}
П> foldMap1 Sum $ unsafeFromList [1]
Sum {getSum = ^C^CInterrupted.
```

You can see this in the implementation for `foldMap1` for NEL: 

```haskell
instance Foldable1 NonEmpty where 
  foldMap1 f (x :| xs) = go (f x) xs 
    where 
      go y [] = y 
      go y (z : zs) = y <> go (f z) zs
```

Reply:
Good catch

Reply:
Here's a suggestion for the code I'd like to see, with some more optimal unpacking and nullity checks. This way, we can exit early without some unnecessary head/tail calls: 

```haskell
instance Foldable1 NonEmptyVector where
    foldMap1 f (NonEmptyVector v) =
      let
        h = V.unsafeHead v
        t = V.unsafeTail v
      in go (f h) t
      where
        go x xs
          | V.null xs = x
          | otherwise =
            let h = V.unsafeHead xs
                t = V.unsafeTail xs
            in x <> go (f h) t
```

Reply:
I couldn't have written it better, and i think the comments are quite useful for understanding the code. So i put it in as is.